### ISSUE TITLE
Add "publish" style command to run functions and copy from layers/modules directly into the project

### ISSUE THREAD

**Issue by Smef** [nuxt/nuxt#21307]
### Describe the feature

It would be very helpful if we had the ability to develop layers or modules for Nuxt which came with a "publish" type of command which could be run by users. This would execute a command in the package which would copy select files from the package into the actual published app.

An example use-case for a feature like this would be to develop something like a layer, but to then publish the pages and components directly into the app's pages or components directories. By copying the content into these folders it allows developers to more easily see the structure of what has been added, but more importantly start making changes to the published files directly. 

We can overwrite content of layers and modules by making our own components and pages which match the package's, but doing that for a significant number of things requires digging into the package source and manually making your own copies of the files, which can be very time-consuming and error-prone.

This kind of feature is very useful when your package has the expectation that what you're publishing is just a template and will generally require customization. 

An example of this behavior in another platform is with [Laravel Jetstream](https://jetstream.laravel.com/). After installing the Jetstream package you run a publish command to copy the starter template of things like login pages, basic form components, nav menus, and other basic app configurations. The expectation is that you as the developer are going to need to make some significant style, theme, or behavior changes to the starter kit which has been copied into your app by the publish process. Having all of those files there, though, makes it very easy to find things like your login page to start changing your styling and branding.

We'd like to make and publish our own version of something like Jetstream for Nuxt, which would give us a starter-kit to cover things like basic auth, menus, password resets, and some starter database configuration. In our organization we'd be using this as a starting point for a lot of projects for different companies, and they'd each need to have significantly different branding and theming. Having the layer content or select module content actually published into the project would make the process of starting to work on something like this very straightforward.

### Additional information

- [X] Would you be willing to help implement this feature?
- [ ] Could this feature be implemented as a module?

### Final checks

- [X] Read the [contribution guide](https://nuxt.com/docs/community/contribution).
- [X] Check existing [discussions](https://github.com/nuxt/nuxt/discussions) and [issues](https://github.com/nuxt/nuxt/issues).

**Comment by Hebilicious** [comment]
If I understand correctly, you'd have a command like `nuxi eject my-module`, and it would copy the code from `my-module` into the local ./modules folder ?

I love this idea, it would be even better if it could copy the source files instead of copying the dist files (maybe resolve them from the package.json repository field, and fallback to dist if it's not present ?)

Eventually have some kind of option to automatically push the changes to github to make it much easier to contribute to modules ?

**Comment by Smef** [comment]
It wouldn't necessarily be the whole contents. Depending on the module and what the package developer thinks the users would care about it might only be a certain subset of files, such as some components, layouts, or configuration files.

Package/module/layer developers may even want to provide people using the package the option to publish or eject certain contents, such as only components and config.

For the example of a starter kit, it would probably want to publish a bunch of components, a config file for the functionality and features, maybe a tailwind config and tsconfig and a couple of other project files.  Things that might not be published might be something like some auth routes or logic. You could still override it if you wanted to, but the package author might decide that it's not worth publishing that by default.

The expectation in this case would be that you're running through this process on a brand new project and that if you aren't for some reason, you're likely to override something in your project with the module's published content.

[Statamic](https://statamic.com/), a CMS, is another example that does something like this as part of an "install" command that you run. When you install Statamic you also "publish" some assets into your main project src. This publish process creates some folders that you'll need, some example files, and some necessary config files. There are still a bunch of files and logic in the package which don't get published, such as a control panel.

Module developers may want to publish everything, but they may also only want to publish a select number of files.

...
...
...

I wrote all this up and then thought was thinking about what an alternative solution would look like. This could be done with an `npx` command from the package. As an example of this, Tailwind has an `npx tailwindcss init` which does some very similar stuff. It publishes some config files as part of an install process.

I think the main difference between this sort of a command and a bit more robust publish type command would be reducing boilerplate that the package developers need to write.
Some examples of saving boilerplate would be:

- Getting paths to specific standard project folders, like the pages directory
- Built-in handling of cases where common folders don't exist yet (again, like a pages directory)
- Standardized handling of file overwrite protection and warnings
- Built-in handling of streaming read/write source to destination files

Standardization of this process could make it easier for module developers to get their content into the right place, and would prevent people from needing to include additional packages in their modules just to do some standard file management stuff.

**Comment by Hebilicious** [comment]
Right, so basically you're suggesting something that would work like this:

- as a module developer you add to some kind of manifest file/config any files that you want to publish/eject.
- you publish your module, these files are added to npm under some kind of special directory in the published package
- the consumer of your module/layers would then run a command and these files would be copied from node_modules to his local directory 
- nuxt.config.ts would get updated 

A couple of things comes to mind :

For modules that don't publish everything then the local version would probably break, so maybe that shouldn't be allowed.
I really think the idea of ejecting a module into its source code would encourage people to contribute back to modules, but that might not work for modules with advanced build steps. Maybe `nuxt-build-module` could test if the module is ejectable and then the `nuxi eject` command would be allowed/disallowed based on that. 

For layers you made a lot of good points, I'm thinking also that if you have relative dependencies in your layers and you don't publish them (ComponentA uses ComponentB and ComponentC, but component C isn't ejected), what should happen when the user ejects?

**Comment by danielroe** [comment]
I think this would be a great idea!

First steps would be to implement this in a standalone package - either/both:

1) a CLI command (`nuxt-eject` is available) - this will be automatically picked up by nuxi to allow you to run `nuxi eject` and copy layers into your project. I think we might also consider supporting auto-resolution of any files in `~/layers` (cc: @pi0) as layers - though note that this would likely require integration with `c12`.
2) a Nuxt module that exposes an iframe to Nuxt Devtools to allow ejecting via Devtools command

Feel free to ping any of us to help.

**Comment by Smef** [comment]
I think we're going to be working on a starter kit for our company to use to bootstrap projects quickly which we plan to open-source, and this would be a good actual test case for this concept.
